text/00110-compatibility-strategy.md

```diff
@@ -0,0 +1,364 @@
+
+An important note is that when testing compatibility to older versions (be it CLI or Framework), we cannot run the *latest* integration test suite, as it may contain tests for new features and bug fixes, which were not implement in older versions.
+
+Therefore, the entire design relies on us having the ability to:
+
+- *Run **old** integration tests using a **new** CLI version and an **old** Framework version.*
+- *Run **old** integration tests using a **new** Framework version and an **old** CLI version.*
+- *Mark a Framework version as incompatible*
+
+    - The tests should support an ability to specify versions who are not compatible, since the expected output would be different.
+
+# Detailed Design
+
+As mentioned, everything is based on our ability to run integration tests with various version permutations. We introduce a new function, `run_integration_tests_perm` which magically does what we need, as well as some useful constants.
+
+The design is best explained by simply looking at the proposed code changes. Most of what you need to know is in the inline comments of the code.
+
+**[common.bash](https://github.com/aws/aws-cdk/blob/master/packages/aws-cdk/test/integ/cli/common.bash)**
+
+```bash
+function bump_minor() {
+
+    version=$1
+
+    # Accept a version and bump the minor version.
+}
+
+function decrease_minor() {
+
+    version=$1
+
+    # Accept a version and decrease the minor version.
+
+}
+
+# This is the latest published version.
+CURRENT_VERSION=$(node -e 'console.log(require("package.json").version')
+
+# This is the soon to be published version.
+PREVIOUS_VERSION="$(decrease_minor ${CURRENT_VERSION})"
+
+# This is the previously published version.
+NEXT_VERSION="$(bump_minor ${CURRENT_VERSION})"
+
+function run_integration_tests_perm() {
+
+    integration_tests_version=1
+    framework_version=$2
+    cli_version=$3
+    expected_error=$4
+
+    # Checkout the relevant integration tests
+    # Install framework version
+    # Install CLI version
+    # Inject versions to integration tests
+    # Run tests (should either pass or print the expected error)
+
+}
+
+```
+
+We now create a new file, which is responsible for validating the compatibility of new framework versions against old cli versions. This test will have the following assertions:
+
+- Compatible versions must pass all integration tests.
+- Incompatible versions must print the expected error message.
+
+### Things to note
+
+- The expected error message in the case of incompatibility is generic. It should not be determined by the developer. This will ensure consistency and correctness.
+- When a version is deemed incompatible, this test needs to be amended with another `if` clause that puts that version in a different flow. We call this process, along with actually bumping the version, `mark-incompatible`. Remember this phrase, as you will see it again a bit further down.
```

Comment:
As I noted earlier, I dislike having to update the test each time a version bump is introduced.